Cross compilation and running of tool and test not working robustly across platforms

Hi there, thank you for this effort, I would like to use lmdb with zig, cross compile and test.
I'm on Debian 12, with zig 0.13.0.

Follows the output

```
$ zig build tools -Dtarget=x86_64-macos
tools
└─ install mplay
   └─ zig build-exe mplay Debug x86_64-macos
      └─ zig build-lib lmdb Debug x86_64-macos 1 errors
/home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb/mdb.c:160:10: error: 'resolv.h' file not found
#include <resolv.h> /* defines BYTE_ORDER on HPUX and Solaris */
         ^~~~~~~~~~~
error: warning(compilation): failed to delete '/home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/tmp/4ea43460777dd969-mdb.o.d': FileNotFound

error: the following command failed with 1 compilation errors:
/home/igor/playground/zig/zig-linux-x86_64-0.13.0/zig build-lib -fno-llvm -fno-lld -cflags -pthread -std=c23 -- /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb/mdb.c /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb/midl.c -fno-strip -fno-sanitize-c -ODebug -target x86_64-macos -mcpu baseline -I /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb -D_XOPEN_SOURCE=600 -Mroot -lc --cache-dir /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache --global-cache-dir /home/igor/.cache/zig --name lmdb -static -fno-lto --listen=- 
Build Summary: 1/15 steps succeeded; 1 failed (disable with --summary none)
tools transitive failure
├─ install mdb_copy transitive failure
│  └─ zig build-exe mdb_copy Debug x86_64-macos transitive failure
│     └─ zig build-lib lmdb Debug x86_64-macos 1 errors
├─ install mdb_drop transitive failure
│  └─ zig build-exe mdb_drop Debug x86_64-macos transitive failure
│     └─ zig build-lib lmdb Debug x86_64-macos (reused)
├─ install mdb_dump transitive failure
│  └─ zig build-exe mdb_dump Debug x86_64-macos transitive failure
│     └─ zig build-lib lmdb Debug x86_64-macos (reused)
├─ install mdb_load transitive failure
│  └─ zig build-exe mdb_load Debug x86_64-macos transitive failure
│     └─ zig build-lib lmdb Debug x86_64-macos (reused)
├─ install mdb_stat transitive failure
│  └─ zig build-exe mdb_stat Debug x86_64-macos transitive failure
│     └─ zig build-lib lmdb Debug x86_64-macos (reused)
└─ install mplay transitive failure
   └─ zig build-exe mplay Debug x86_64-macos transitive failure
      └─ zig build-lib lmdb Debug x86_64-macos (reused)
error: the following build command failed with exit code 1:
/home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/8b8a090cd1cdbaa98919363df8845d5e/build /home/igor/playground/zig/zig-linux-x86_64-0.13.0/zig /home/igor/playground/zig/lmdb-options/acbtu /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache /home/igor/.cache/zig --seed 0x83b9d081 -Z83ccb1a88b79bbb2 tools -Dtarget=x86_64-macos
```

But I'm afraid there is a regression for Linux on tools I'm gettting:

```
$ zig build tools -Dtarget=x86_64-linux --verbose
/home/igor/playground/zig/zig-linux-x86_64-0.13.0/zig build-lib -fno-llvm -fno-lld -cflags -pthread -std=c23 -- /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb/mdb.c /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb/midl.c -fno-strip -fno-sanitize-c -ODebug -target x86_64-linux -mcpu baseline -I /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb -D_XOPEN_SOURCE=600 -Mroot -lc --cache-dir /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache --global-cache-dir /home/igor/.cache/zig --name lmdb -static -fno-lto --listen=- 
/home/igor/playground/zig/zig-linux-x86_64-0.13.0/zig build-exe -fno-llvm -fno-lld -cflags -pthread -std=c23 -- /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb/mdb_dump.c /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/c77a71d36d36dcc44242e5a83184ee50/liblmdb.a -fno-strip -fno-sanitize-c -ODebug -target x86_64-linux -mcpu baseline -I /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb -I /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/d926581394ea455c2cfb25b410d07bd3 -D_XOPEN_SOURCE=600 -Mroot -lc --cache-dir /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache --global-cache-dir /home/igor/.cache/zig --name mdb_dump --listen=- 
/home/igor/playground/zig/zig-linux-x86_64-0.13.0/zig build-exe -fno-llvm -fno-lld -cflags -pthread -std=c23 -- /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb/mdb_stat.c /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/c77a71d36d36dcc44242e5a83184ee50/liblmdb.a -fno-strip -fno-sanitize-c -ODebug -target x86_64-linux -mcpu baseline -I /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb -I /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/d926581394ea455c2cfb25b410d07bd3 -D_XOPEN_SOURCE=600 -Mroot -lc --cache-dir /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache --global-cache-dir /home/igor/.cache/zig --name mdb_stat --listen=- 
/home/igor/playground/zig/zig-linux-x86_64-0.13.0/zig build-exe -fno-llvm -fno-lld -cflags -pthread -std=c23 -- /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb/mdb_load.c /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/c77a71d36d36dcc44242e5a83184ee50/liblmdb.a -fno-strip -fno-sanitize-c -ODebug -target x86_64-linux -mcpu baseline -I /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb -I /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/d926581394ea455c2cfb25b410d07bd3 -D_XOPEN_SOURCE=600 -Mroot -lc --cache-dir /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache --global-cache-dir /home/igor/.cache/zig --name mdb_load --listen=- 
/home/igor/playground/zig/zig-linux-x86_64-0.13.0/zig build-exe -fno-llvm -fno-lld -cflags -pthread -std=c23 -- /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb/mdb_copy.c /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/c77a71d36d36dcc44242e5a83184ee50/liblmdb.a -fno-strip -fno-sanitize-c -ODebug -target x86_64-linux -mcpu baseline -I /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb -I /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/d926581394ea455c2cfb25b410d07bd3 -D_XOPEN_SOURCE=600 -Mroot -lc --cache-dir /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache --global-cache-dir /home/igor/.cache/zig --name mdb_copy --listen=- 
/home/igor/playground/zig/zig-linux-x86_64-0.13.0/zig build-exe -fno-llvm -fno-lld -cflags -pthread -std=c23 -- /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb/mplay.c /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/c77a71d36d36dcc44242e5a83184ee50/liblmdb.a -fno-strip -fno-sanitize-c -ODebug -target x86_64-linux -mcpu baseline -I /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb -I /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/d926581394ea455c2cfb25b410d07bd3 -D_XOPEN_SOURCE=600 -Mroot -lc --cache-dir /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache --global-cache-dir /home/igor/.cache/zig --name mplay --listen=- 
/home/igor/playground/zig/zig-linux-x86_64-0.13.0/zig build-exe -fno-llvm -fno-lld -cflags -pthread -std=c23 -- /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb/mdb_drop.c /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/c77a71d36d36dcc44242e5a83184ee50/liblmdb.a -fno-strip -fno-sanitize-c -ODebug -target x86_64-linux -mcpu baseline -I /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb -I /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/d926581394ea455c2cfb25b410d07bd3 -D_XOPEN_SOURCE=600 -Mroot -lc --cache-dir /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache --global-cache-dir /home/igor/.cache/zig --name mdb_drop --listen=- 
tools
└─ install mplay
   └─ zig build-exe mplay Debug x86_64-linux 2 errors
/home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb/mplay.c:488:9: error: cannot assign to variable 'stdin' with const-qualified type 'FILE *const' (aka 'struct _IO_FILE *const')
  stdin = fdopen(0, "r");
  ~~~~~~^
/home/igor/playground/zig/zig-linux-x86_64-0.13.0/lib/libc/include/generic-musl/stdio.h:62:20: note: variable 'stdin' declared const here
extern FILE *const stdin;
~~~~~~~~~~~~~~~~~~~^
/home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb/mplay.c:489:10: error: cannot assign to variable 'stdout' with const-qualified type 'FILE *const' (aka 'struct _IO_FILE *const')
  stdout = fdopen(1, "w");
  ~~~~~~~^
/home/igor/playground/zig/zig-linux-x86_64-0.13.0/lib/libc/include/generic-musl/stdio.h:63:20: note: variable 'stdout' declared const here
extern FILE *const stdout;
~~~~~~~~~~~~~~~~~~~^
error: the following command failed with 2 compilation errors:
/home/igor/playground/zig/zig-linux-x86_64-0.13.0/zig build-exe -fno-llvm -fno-lld -cflags -pthread -std=c23 -- /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb/mplay.c /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/c77a71d36d36dcc44242e5a83184ee50/liblmdb.a -fno-strip -fno-sanitize-c -ODebug -target x86_64-linux -mcpu baseline -I /home/igor/.cache/zig/p/12205c0f7d4ad44b671cf89a7793aafa31ba4739f01b706313e4eadc944b734bd224/libraries/liblmdb -I /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/d926581394ea455c2cfb25b410d07bd3 -D_XOPEN_SOURCE=600 -Mroot -lc --cache-dir /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache --global-cache-dir /home/igor/.cache/zig --name mplay --listen=- 
Build Summary: 12/15 steps succeeded; 1 failed (disable with --summary none)
tools transitive failure
└─ install mplay transitive failure
   └─ zig build-exe mplay Debug x86_64-linux 2 errors
error: the following build command failed with exit code 1:
/home/igor/playground/zig/lmdb-options/acbtu/.zig-cache/o/1a321326e5945f148e274f70a9343cff/build /home/igor/playground/zig/zig-linux-x86_64-0.13.0/zig /home/igor/playground/zig/lmdb-options/acbtu /home/igor/playground/zig/lmdb-options/acbtu/.zig-cache /home/igor/.cache/zig --seed 0xdef3745c -Zfb25c20541125046 tools -Dtarget=x86_64-linux --verbose
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -62,8 +62,10 @@ jobs:
           version: ${{ matrix.zig-version }}
           use-cache: false
 
-      - name: Run `build tools on macos`
-        run: zig build tools -Dtarget=x86_64-macos-none --summary all
+      - name: Run `build tools on macos and windows`
+        run: |
+          zig build tools -Dtarget=x86_64-macos --summary all
+          zig build tools -Dtarget=x86_64-windows --summary all
 
   test:
     strategy:

--- a/build.zig
+++ b/build.zig
@@ -124,7 +124,6 @@ const BuildLmdb = struct {
     }
 
     fn tools(bl: BuildLmdb, liblmdb: *Compile) void {
-        const opt = bl.opt;
         const b = bl.b;
         const tools_step = b.step("tools", "Install lmdb tools");
 
@@ -170,11 +169,14 @@ const BuildLmdb = struct {
             "mdb_load.c",
             "mdb_stat.c",
         };
-        // windows doesn't have 'sys/wait.h' its a posix system header
-        const extra_tools = [_][]const u8{"mplay.c"};
-
         build_tools(bl, liblmdb, tools_step, core_tools[0..]);
-        if (!opt.isWindows()) build_tools(bl, liblmdb, tools_step, extra_tools[0..]);
+
+        // Disable mplay because windows doesn't have the posix system header
+        // 'sys/wait.h' and it doesn't compile on `musl` libc becaues `stdin`
+        // and `stdout` are defined as `const` which `mplay.c` tries to modify
+        const extra_tools = [_][]const u8{"mplay.c"};
+        const disable_mplay = true;
+        if (!disable_mplay) build_tools(bl, liblmdb, tools_step, extra_tools[0..]);
     }
 
     fn tests(bl: BuildLmdb, liblmdb: *Compile) void {

--- a/build.zig
+++ b/build.zig
@@ -1,8 +1,21 @@
 const std = @import("std");
-const builtin = @import("builtin");
 const mem = std.mem;
 const Build = std.Build;
 const Step = Build.Step;
+const Compile = Step.Compile;
+const builtin = @import("builtin");
+
+const MakeOptions = if (@hasDecl(Step, "MakeOptions"))
+    Step.MakeOptions
+else
+    std.Progress.Node;
+
+const lmdb_root = "libraries/liblmdb";
+
+const cflags = .{
+    "-pthread",
+    "-std=c23",
+};
 
 pub fn build(b: *Build) void {
     if (comptime !checkVersion())
@@ -11,243 +24,305 @@ pub fn build(b: *Build) void {
     const target = b.standardTargetOptions(.{});
     const optimize = b.standardOptimizeOption(.{});
 
+    const strip = b.option(bool, "strip", "Strip debug information") orelse false;
+    const lto = b.option(bool, "lto", "Enable link time optimization") orelse false;
+
     const lmdb_upstream = b.dependency(
         "lmdb",
         .{ .target = target, .optimize = optimize },
     );
-    const lmdb_root = "libraries/liblmdb";
 
-    const strip = b.option(bool, "strip", "Strip debug information") orelse false;
-    const lto = b.option(bool, "lto", "Enable link time optimization") orelse false;
-
-    const is_macos = builtin.os.tag == .macos;
-    const need_resolve_h = is_macos or target.result.os.tag == .macos;
-    const is_windows = builtin.os.tag == .windows;
-
-    // writing WritingLibFiles in zld isn't implemented on windows
-    // and zld is the only linker supported on macos
-    const use_lld = if (is_macos) false else if (is_windows) true else switch (optimize) {
-        .Debug => false,
-        else => true,
-    };
-    const use_lto = if (is_macos) false else if (use_lld) lto else false;
-
-    const liblmdb = b.addStaticLibrary(.{
-        .name = "lmdb",
+    const build_opt: BuildOpt = .{
+        .lmdb_upstream = lmdb_upstream,
+        .lto_option = lto,
+        .strip_option = strip,
         .target = target,
         .optimize = optimize,
-        .link_libc = true,
-        .strip = strip,
-        .use_llvm = switch (optimize) {
-            .Debug => if (is_windows) true else false,
-            else => true,
-        },
-        .use_lld = use_lld,
-    });
-    liblmdb.want_lto = use_lto;
-    liblmdb.root_module.sanitize_c = false;
-
-    const liblmdb_src = .{
-        "mdb.c",
-        "midl.c",
     };
-    const lmdb_includes = .{
-        "lmdb.h",
-        "midl.h",
-    };
-    const cflags = .{
-        "-pthread",
-        "-std=c23",
+    const lmdb: BuildLmdb = .{
+        .b = b,
+        .opt = build_opt,
     };
 
-    liblmdb.addCSourceFiles(.{
-        .root = lmdb_upstream.path(lmdb_root),
-        .files = &liblmdb_src,
-        .flags = &cflags,
-    });
-    liblmdb.addIncludePath(lmdb_upstream.path(lmdb_root));
-    liblmdb.root_module.addCMacro("_XOPEN_SOURCE", "600");
-    if (need_resolve_h) {
-        liblmdb.root_module.addCMacro("_DARWIN_C_SOURCE", "");
-    }
+    const liblmdb = lmdb.lib();
+    lmdb.tools(liblmdb);
+    lmdb.api();
+    lmdb.tests(liblmdb);
+}
 
-    liblmdb.installHeadersDirectory(
-        lmdb_upstream.path(lmdb_root),
-        "",
-        .{ .include_extensions = &lmdb_includes },
-    );
+const BuildLmdb = struct {
+    b: *Build,
+    opt: BuildOpt,
 
-    b.installArtifact(liblmdb);
-
-    const lmdb_tools = [_][]const u8{
-        "mdb_copy.c",
-        "mdb_drop.c",
-        "mdb_dump.c",
-        "mdb_load.c",
-        "mdb_stat.c",
-        "mplay.c",
-    };
-
-    const tools_step = b.step("tools", "Install lmdb tools");
-
-    for (lmdb_tools) |tool_file| {
-        const bin_name = tool_file[0..mem.indexOfScalar(u8, tool_file, '.').?];
-        const tool = b.addExecutable(.{
-            .name = bin_name,
-            .target = target,
-            .optimize = optimize,
+    fn lib(bl: BuildLmdb) *Compile {
+        const opt = bl.opt;
+        const b = bl.b;
+        const liblmdb = b.addStaticLibrary(.{
+            .name = "lmdb",
+            .target = opt.target,
+            .optimize = opt.optimize,
             .link_libc = true,
-            .strip = strip,
-            .use_llvm = switch (optimize) {
-                .Debug => false,
-                else => true,
-            },
-            .use_lld = use_lld,
+            .strip = opt.strip_option,
+            .use_llvm = opt.use_llvm(),
+            .use_lld = opt.use_lld(),
         });
-        tool.root_module.sanitize_c = false;
+        liblmdb.want_lto = opt.use_lto();
+        liblmdb.root_module.sanitize_c = false;
 
-        tool.addCSourceFiles(.{
-            .root = lmdb_upstream.path(lmdb_root),
-            .files = &.{tool_file},
+        const lmdb_includes = .{
+            "lmdb.h",
+            "midl.h",
+        };
+        const liblmdb_src = .{
+            "mdb.c",
+            "midl.c",
+        };
+
+        liblmdb.addCSourceFiles(.{
+            .root = opt.lmdb_upstream.path(lmdb_root),
+            .files = &liblmdb_src,
             .flags = &cflags,
         });
-        tool.addIncludePath(lmdb_upstream.path(lmdb_root));
-        tool.root_module.addCMacro("_XOPEN_SOURCE", "600");
-        if (need_resolve_h) {
-            tool.root_module.addCMacro("_DARWIN_C_SOURCE", "");
+        liblmdb.addIncludePath(opt.lmdb_upstream.path(lmdb_root));
+        liblmdb.root_module.addCMacro("_XOPEN_SOURCE", "600");
+        if (opt.isMacos()) {
+            liblmdb.root_module.addCMacro("_DARWIN_C_SOURCE", "");
         }
-        tool.linkLibrary(liblmdb);
 
-        const install_tool = b.addInstallArtifact(tool, .{});
-        tools_step.dependOn(&install_tool.step);
+        liblmdb.installHeadersDirectory(
+            opt.lmdb_upstream.path(lmdb_root),
+            "",
+            .{ .include_extensions = &lmdb_includes },
+        );
+        b.installArtifact(liblmdb);
+
+        return liblmdb;
     }
 
-    const lmdb_api = b.addTranslateC(.{
-        .root_source_file = b.path("include/c.h"),
-        .target = b.graph.host,
-        .optimize = .Debug,
-    });
+    fn api(bl: BuildLmdb) void {
+        const b = bl.b;
+        const opt = bl.opt;
+        const lmdb_api = b.addTranslateC(.{
+            .root_source_file = b.path("include/c.h"),
+            .target = b.graph.host,
+            .optimize = .Debug,
+        });
 
-    if (@hasDecl(Step.TranslateC, "addIncludeDir")) {
-        const path = lmdb_upstream.path(lmdb_root);
-        const absolute_include = path.getPath2(b, null);
-        lmdb_api.addIncludeDir(absolute_include);
-    } else {
-        lmdb_api.addIncludePath(lmdb_upstream.path(lmdb_root));
+        if (@hasDecl(Step.TranslateC, "addIncludeDir")) {
+            const path = opt.lmdb_upstream.path(lmdb_root);
+            const absolute_include = path.getPath2(b, null);
+            lmdb_api.addIncludeDir(absolute_include);
+        } else {
+            lmdb_api.addIncludePath(opt.lmdb_upstream.path(lmdb_root));
+        }
+
+        _ = b.addModule("lmdb", .{
+            .root_source_file = lmdb_api.getOutput(),
+            .target = opt.target,
+            .optimize = opt.optimize,
+        });
     }
 
-    _ = b.addModule("lmdb", .{
-        .root_source_file = lmdb_api.getOutput(),
-        .target = target,
-        .optimize = optimize,
-    });
+    fn tools(bl: BuildLmdb, liblmdb: *Compile) void {
+        const opt = bl.opt;
+        const b = bl.b;
+        const tools_step = b.step("tools", "Install lmdb tools");
 
-    const cflags_test = .{
-        "-pthread",
-        "-std=c17", //c23 forbids function use without prototype
-        "-Wno-format",
-        "-Wno-implicit-function-declaration",
-    };
+        const build_tools = struct {
+            fn build_tools(bl_: BuildLmdb, liblmdb_: *Compile, tools_step_: *Step, lmdb_tools: []const []const u8) void {
+                const b_ = bl_.b;
+                const opt_ = bl_.opt;
+                for (lmdb_tools) |tool_file| {
+                    const bin_name = tool_file[0..mem.indexOfScalar(u8, tool_file, '.').?];
+                    const tool = b_.addExecutable(.{
+                        .name = bin_name,
+                        .target = opt_.target,
+                        .optimize = opt_.optimize,
+                        .link_libc = true,
+                        .strip = opt_.strip_option,
+                        .use_llvm = opt_.use_llvm(),
+                        .use_lld = opt_.use_lld(),
+                    });
+                    tool.root_module.sanitize_c = false;
 
-    const lmdb_test = [_][]const u8{
-        "mtest.c",
-        "mtest2.c",
-        "mtest3.c",
-        "mtest4.c",
-        "mtest5.c",
-        // "mtest6.c", // disabled as it requires building liblmdb with MDB_DEBUG
-    };
+                    tool.addCSourceFiles(.{
+                        .root = opt_.lmdb_upstream.path(lmdb_root),
+                        .files = &.{tool_file},
+                        .flags = &cflags,
+                    });
+                    tool.addIncludePath(opt_.lmdb_upstream.path(lmdb_root));
+                    tool.root_module.addCMacro("_XOPEN_SOURCE", "600");
+                    if (opt_.isMacos()) {
+                        tool.root_module.addCMacro("_DARWIN_C_SOURCE", "");
+                    }
+                    tool.linkLibrary(liblmdb_);
 
-    const MakeOptions = if (@hasDecl(Step, "MakeOptions"))
-        Step.MakeOptions
-    else
-        std.Progress.Node;
+                    const install_tool = b_.addInstallArtifact(tool, .{});
+                    tools_step_.dependOn(&install_tool.step);
+                }
+            }
+        }.build_tools;
 
-    const test_step = b.step("test", "Run lmdb tests");
-    const run_create_testdb = struct {
-        fn makeFn(step: *Step, options: MakeOptions) !void {
-            _ = options;
-            const test_run = Step.cast(step, Step.Run).?;
-            const subpath = "testdb/";
-            if (@hasDecl(Build.LazyPath, "getPath3")) {
-                const bin_path = test_run.cwd.?.getPath3(step.owner, step);
-                bin_path.makePath(subpath) catch unreachable;
-            } else {
-                const bin_path = test_run.cwd.?.getPath2(step.owner, step);
-                const owner = test_run.step.owner;
-                const full_path = owner.fmt("{s}/{s}", .{ bin_path, subpath });
-                owner.cache_root.handle.makeDir(full_path) catch |err| switch (err) {
+        const core_tools = [_][]const u8{
+            "mdb_copy.c",
+            "mdb_drop.c",
+            "mdb_dump.c",
+            "mdb_load.c",
+            "mdb_stat.c",
+        };
+        // windows doesn't have 'sys/wait.h' its a posix system header
+        const extra_tools = [_][]const u8{"mplay.c"};
+
+        build_tools(bl, liblmdb, tools_step, core_tools[0..]);
+        if (!opt.isWindows()) build_tools(bl, liblmdb, tools_step, extra_tools[0..]);
+    }
+
+    fn tests(bl: BuildLmdb, liblmdb: *Compile) void {
+        const b = bl.b;
+        const opt = bl.opt;
+        const install_test_step = b.step("install-test", "Install lmdb tests");
+
+        const install_test_subpath = "test/";
+        install_test_step.makeFn = struct {
+            fn makeFn(step: *Step, options: MakeOptions) !void {
+                _ = options;
+                const step_build = step.owner;
+                std.fs.cwd().makeDir(step_build.fmt(
+                    "{s}/{s}/testdb/",
+                    .{ step_build.install_prefix, install_test_subpath },
+                )) catch |err| switch (err) {
                     error.PathAlreadyExists => {},
                     else => unreachable,
                 };
             }
-        }
+        }.makeFn;
 
-        fn create_testdb(owner: *Build, test_dirname: Build.LazyPath) *Step {
-            const run = Step.Run.create(owner, "create testdb at the generated path");
-            run.step.makeFn = makeFn;
+        const create_testdb = struct {
+            fn makeFn(step: *Step, options: MakeOptions) !void {
+                _ = options;
+                const test_run = Step.cast(step, Step.Run).?;
+                const subpath = "testdb/";
+                if (@hasDecl(Build.LazyPath, "getPath3")) {
+                    const bin_path = test_run.cwd.?.getPath3(step.owner, step);
+                    bin_path.makePath(subpath) catch unreachable;
+                } else {
+                    const bin_path = test_run.cwd.?.getPath2(step.owner, step);
+                    const owner = test_run.step.owner;
+                    const full_path = owner.fmt("{s}/{s}", .{ bin_path, subpath });
+                    owner.cache_root.handle.makeDir(full_path) catch |err| switch (err) {
+                        error.PathAlreadyExists => {},
+                        else => unreachable,
+                    };
+                }
+            }
+
+            fn create_testdb(owner: *Build, test_dirname: Build.LazyPath) *Step {
+                const run = Step.Run.create(owner, "create testdb at the generated path");
+                run.step.makeFn = makeFn;
+                run.setCwd(test_dirname);
+
+                return &run.step;
+            }
+        }.create_testdb;
+
+        const test_step = b.step("test", "Run lmdb tests");
+
+        const lmdb_test = [_][]const u8{
+            "mtest.c",
+            "mtest2.c",
+            "mtest3.c",
+            "mtest4.c",
+            "mtest5.c",
+            // "mtest6.c", // disabled as it requires building liblmdb with MDB_DEBUG
+        };
+
+        const cflags_test = .{
+            "-pthread",
+            "-std=c17", //c23 forbids function use without prototype
+            "-Wno-format",
+            "-Wno-implicit-function-declaration",
+        };
+
+        for (lmdb_test) |test_file| {
+            const test_name = test_file[0..mem.indexOfScalar(u8, test_file, '.').?];
+
+            const test_exe = b.addExecutable(.{
+                .name = test_name,
+                .target = opt.target,
+                .optimize = .Debug,
+                .link_libc = true,
+                .use_lld = opt.use_lld(),
+            });
+            test_exe.root_module.sanitize_c = false;
+
+            test_exe.addCSourceFiles(.{
+                .root = opt.lmdb_upstream.path(lmdb_root),
+                .files = &.{test_file},
+                .flags = &cflags_test,
+            });
+            test_exe.addIncludePath(opt.lmdb_upstream.path(lmdb_root));
+            test_exe.linkLibrary(liblmdb);
+
+            const test_dirname = test_exe.getEmittedBin().dirname();
+
+            const install_test_exe = b.addInstallArtifact(test_exe, .{ .dest_dir = .{ .override = .{
+                .custom = install_test_subpath,
+            } } });
+
+            const run = b.addRunArtifact(test_exe);
             run.setCwd(test_dirname);
+            run.expectExitCode(0);
+            run.enableTestRunnerMode();
 
-            return &run.step;
+            const run_create_testdb = create_testdb(run.step.owner, test_dirname);
+            run_create_testdb.dependOn(&test_exe.step);
+            run.step.dependOn(run_create_testdb);
+            test_step.dependOn(&run.step);
+
+            install_test_step.dependOn(&install_test_exe.step);
         }
-    }.create_testdb;
-
-    const install_test_step = b.step("install-test", "Install lmdb tests");
-
-    const install_test_subpath = "test/";
-    install_test_step.makeFn = struct {
-        fn makeFn(step: *Step, options: MakeOptions) !void {
-            _ = options;
-            const step_build = step.owner;
-            std.fs.cwd().makeDir(step_build.fmt(
-                "{s}/{s}/testdb/",
-                .{ step_build.install_prefix, install_test_subpath },
-            )) catch |err| switch (err) {
-                error.PathAlreadyExists => {},
-                else => unreachable,
-            };
-        }
-    }.makeFn;
-
-    for (lmdb_test) |test_file| {
-        const test_name = test_file[0..mem.indexOfScalar(u8, test_file, '.').?];
-
-        const test_exe = b.addExecutable(.{
-            .name = test_name,
-            .target = target,
-            .optimize = .Debug,
-            .link_libc = true,
-            .use_lld = use_lld,
-        });
-        test_exe.root_module.sanitize_c = false;
-
-        test_exe.addCSourceFiles(.{
-            .root = lmdb_upstream.path(lmdb_root),
-            .files = &.{test_file},
-            .flags = &cflags_test,
-        });
-        test_exe.addIncludePath(lmdb_upstream.path(lmdb_root));
-        test_exe.linkLibrary(liblmdb);
-
-        const test_dirname = test_exe.getEmittedBin().dirname();
-
-        const install_test_exe = b.addInstallArtifact(test_exe, .{ .dest_dir = .{ .override = .{
-            .custom = install_test_subpath,
-        } } });
-
-        const run = b.addRunArtifact(test_exe);
-        run.setCwd(test_dirname);
-        run.expectExitCode(0);
-        run.enableTestRunnerMode();
-
-        const create_testdb = run_create_testdb(run.step.owner, test_dirname);
-        create_testdb.dependOn(&test_exe.step);
-        run.step.dependOn(create_testdb);
-        test_step.dependOn(&run.step);
-
-        install_test_step.dependOn(&install_test_exe.step);
     }
-}
+};
+
+const BuildOpt = struct {
+    lmdb_upstream: *Build.Dependency,
+    target: Build.ResolvedTarget,
+    optimize: std.builtin.OptimizeMode,
+    strip_option: bool,
+    lto_option: bool,
+
+    fn isOs(os: std.Target.Os.Tag, target: Build.ResolvedTarget) bool {
+        return builtin.os.tag == os or target.result.os.tag == os;
+    }
+
+    fn isMacos(opt: BuildOpt) bool {
+        return isOs(.macos, opt.target);
+    }
+
+    fn isWindows(opt: BuildOpt) bool {
+        return isOs(.windows, opt.target);
+    }
+
+    fn use_lto(opt: BuildOpt) bool {
+        return if (opt.isMacos()) false else if (opt.use_lld()) opt.lto_option else false;
+    }
+
+    fn use_llvm(opt: BuildOpt) bool {
+        return switch (opt.optimize) {
+            .Debug => if (opt.isWindows()) true else false,
+            else => true,
+        };
+    }
+
+    // writing WritingLibFiles in zld isn't implemented on windows
+    // and zld is the only linker supported on macos
+    fn use_lld(opt: BuildOpt) bool {
+        return if (opt.isMacos()) false else if (opt.isWindows()) true else switch (opt.optimize) {
+            .Debug => false,
+            else => true,
+        };
+    }
+};
 
 // ensures the currently in-use zig version is at least the minimum required
 fn checkVersion() bool {

